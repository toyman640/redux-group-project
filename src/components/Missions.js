import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchMissions, joinMission, leaveMission } from '../redux/missions/missionsSlice';
import '../App.css';

const Missions = () => {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);
  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);
  const handleMission = (missionId, currentStatus) => {
    if (currentStatus) {
      dispatch(leaveMission(missionId));
    } else {
      dispatch(joinMission(missionId));
    }
  };
  return (
    <div>
      <table className="Mission-table">
        <thead>
          <tr className="Table-head">
            <th className="Mission-title">Mission</th>
            <th className="Desc-title">Discription</th>
            <th className="Desc-title">Status</th>
            <th className="Desc-title-2"> </th>
          </tr>
        </thead>
        {missions.map((mission, index) => (
          <tbody key={mission.id}>
            <tr className={index % 2 === 1 ? 'Even-row' : 'Odd-row'}>
              <td className="Mission-name">{mission.name}</td>
              <td className="Description">{mission.description}</td>
              <td className="Status">
                <button type="button" className={mission.reserved ? 'Active-member' : 'Not-member'}>
                  {mission.reserved ? 'Active Member' : 'Not a Member'}
                </button>
              </td>
              <td className="Status-2">
                <button type="button" onClick={() => handleMission(mission.id, mission.reserved)} className={mission.reserved ? 'Leave-mission' : 'Join-mission'}>
                  {mission.reserved ? 'Leave Mission' : 'Join Mission'}
                </button>
              </td>
            </tr>
          </tbody>
        ))}
      </table>
    </div>
  );
};

export default Missions;
