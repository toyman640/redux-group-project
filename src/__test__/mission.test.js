import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { useSelector, useDispatch } from 'react-redux';
import Missions from '../components/Missions';

jest.mock('react-redux');

describe('A test for Missions component', () => {
  it('renders correctly', () => {
    const missions = [
      {
        id: 1, name: 'Mission 1', description: 'Description 1', reserved: true,
      },
      {
        id: 2, name: 'Mission 2', description: 'Description 2', reserved: false,
      },
    ];

    useSelector.mockImplementation((selectorFn) => selectorFn({
      missions: { missions },
    }));

    const dispatchMock = jest.fn();
    useDispatch.mockReturnValue(dispatchMock);

    const tree = renderer.create(<Missions />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders mission table with correct data', () => {
    const missions = [
      {
        id: 1, name: 'Mission 1', description: 'Description 1', reserved: true,
      },
      {
        id: 2, name: 'Mission 2', description: 'Description 2', reserved: false,
      },
    ];

    useSelector.mockImplementation((selectorFn) => selectorFn({
      missions: { missions },
    }));

    const dispatchMock = jest.fn();
    useDispatch.mockReturnValue(dispatchMock);

    const { getByText } = render(<Missions />);

    missions.forEach((mission) => {
      expect(getByText(mission.name)).toBeInTheDocument();
      expect(getByText(mission.description)).toBeInTheDocument();

      const statusButton = getByText(mission.reserved ? 'Active Member' : 'Not a Member');
      expect(statusButton).toBeInTheDocument();
      expect(statusButton).toHaveClass(mission.reserved ? 'Active-member' : 'Not-member');

      const actionButton = getByText(mission.reserved ? 'Leave Mission' : 'Join Mission');
      expect(actionButton).toBeInTheDocument();
      expect(actionButton).toHaveClass(mission.reserved ? 'Leave-mission' : 'Join-mission');
    });
  });
});
