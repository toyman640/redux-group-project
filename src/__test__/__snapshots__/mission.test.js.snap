// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A test for Missions component renders correctly 1`] = `
<div>
  <table
    className="Mission-table"
  >
    <thead>
      <tr
        className="Table-head"
      >
        <th
          className="Mission-title"
        >
          Mission
        </th>
        <th
          className="Desc-title"
        >
          Discription
        </th>
        <th
          className="Desc-title"
        >
          Status
        </th>
        <th
          className="Desc-title-2"
        >
           
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="Odd-row"
      >
        <td
          className="Mission-name"
        >
          Mission 1
        </td>
        <td
          className="Description"
        >
          Description 1
        </td>
        <td
          className="Status"
        >
          <button
            className="Active-member"
            type="button"
          >
            Active Member
          </button>
        </td>
        <td
          className="Status-2"
        >
          <button
            className="Leave-mission"
            onClick={[Function]}
            type="button"
          >
            Leave Mission
          </button>
        </td>
      </tr>
    </tbody>
    <tbody>
      <tr
        className="Even-row"
      >
        <td
          className="Mission-name"
        >
          Mission 2
        </td>
        <td
          className="Description"
        >
          Description 2
        </td>
        <td
          className="Status"
        >
          <button
            className="Not-member"
            type="button"
          >
            Not a Member
          </button>
        </td>
        <td
          className="Status-2"
        >
          <button
            className="Join-mission"
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
